Dask.config set and get normalize underscores and hyphens

jrbourbeau opened the pull request:
This PR updates `dask.config.get` and `dask.config.set` to both replace single underscores with hyphens. The motivation for this is to avoid a mis-match between configuration variables in Python with underscores, e.g., 

```python
with dask.config.set(num_workers=5):
    ...
```
 
and environment variables, e.g. `DASK_NUM_WORKERS`, which already replace `'_'` with `'-'`.

Fixes #4141

- [x] Tests added / passed
- [x] Passes `flake8 dask`

jrbourbeau commented:
Currently, this PR doesn't handle any user-defined custom keys with underscores. For example, if in a config yaml file, someone has

```
custom_key:
    nested_key1: value1
    nested_key2: value2
```

I'm planning on adding functionality to handle this case.

mrocklin commented on dask/tests/test_base.py:
```diff
@@ -885,3 +885,23 @@ def f(x):
+
+    workers = {i.worker_id for i in prof.results}
+
+    assert len(workers) == num_workers
```
I recommend that we test this functionality more directly.  For example we might set an environment variable that you ahve here, and then verify that the correct value is returned by both `dask.config.get('num_workers')` and dask.config.get('num-workers')`.  We might also use some more generic key, like `a-b` / `a_b`.  It would also be nice to verify that you can use either one as we roundtrip between get/set rather than using environment variables.  

I think that a more direct test will also have the advantage of being faster (this one takes a second or more if I recall correctly) and will also probably communicate the intent of these changes to future developers more clearly.

mrocklin commented:
@djhoese I think that you may find this PR interesting.

djhoese commented:
> @djhoese I think that you may find this PR interesting.

Yes both for the issue and my config object stuff. After looking at these changes, is there a reason that environment variables have to be converted to use hyphens at all? Can't they keep their underscore naming? It seems like most formats are going to allow underscores without issue, but hyphens aren't always allowed. Right?

mrocklin commented:
There are a number of operational situations where you want to use an environment variable to specify a config value that uses hyphens.  Most of the config values use hyphens today.

mrocklin commented:
We could change the use of hyphens to underscores, but that would be a longish deprecation cycle.

I like the idea of being blind to hyphen/underscore differences.

djhoese commented:
I think you misunderstand or maybe I do. This PR is about normalizing *keys* not *values*. Currently all environment variable *keys* are normalized by converting `_` to `-` (https://github.com/dask/dask/blob/master/dask/config.py#L152). If any environment variable *keys* are benefiting from this underscore to hyphen conversion then don't they suffer from the same issues that this PR is trying to solve?

mrocklin commented:
Right, lets say that I want to populate the config value for `scheduler-address` with an environment variable.  Today I do that with 

    DASK_SCHEDULER_ADDRESS='tcp://...'

which converts to 

    {'scheduler-address': 'tcp://...'}

> is there a reason that environment variables have to be converted to use hyphens at all? Can't they keep their underscore naming?

If we went with that policy then the situation above wouldn't be feasible.  There wouldn't be a way to set `scheduler-address` with an environment variable.

mrocklin commented:
The proposal in this PR is to make `scheduler_address` and `scheduler-address` the same, so as to avoid this issue.

djhoese commented:
Ok so currently it is impossible to do `dask.config.set(scheduler_address='tcp://...')`, right?

mrocklin commented:
That wouldn't affect the `scheduler-address` config value, correct.

djhoese commented:
Ok so there is no way to adjust certain config values with `set` currently. If that hasn't been a problem previously then this all makes sense why it is popping up now and why underscores can't be used without causing backwards compatibility issues. Thanks for the explanation.

jrbourbeau commented:
Just added a `normalize_nested_keys` function to handle any keys with underscores in configuration yaml files (ref https://github.com/dask/dask/pull/4143#issuecomment-433530286). 

@mrocklin I also added tests to directly check that `set` and `get` are invariant under using either `'_'` or `'-'`.

jrbourbeau commented on dask/tests/test_config.py:
```diff
@@ -297,3 +298,43 @@ def test_expand_environment_variables(inp, out):
+    assert get('a-b') == value
+
+    # Maks sure to remove this test key from config
+    dask.config.config.pop('a-b')
```
Yeah, that's what I was thinking too. Thanks for the feedback

jrbourbeau replied:
OoOo actually it looks like `get` and `refresh` have `config=` keyword arguments that let you provide your own configuration mapping instead of using the global `dask.config.config`

djhoese commented:
@jrbourbeau FYI I'm working on migrating the dask config *module* to a Config object in this PR: https://github.com/dask/dask/pull/4099

I was going to ask that you try not use the `config` keyword argument, but it actually isn't that hard to work around in the my tests iirc.

jcrist commented on dask/config.py:
```diff
@@ -225,6 +246,17 @@ def ensure_file(
         pass
 
 
+def normalize_key(key):
```
Nit: I'd prefer this was located closer to `normalize_nested_keys`, to help readability.

jrbourbeau replied:
Makes sense, will do

djhoese commented on dask/config.py:
```diff
@@ -102,6 +102,25 @@ def merge(*dicts):
+    Examples
+    --------
+    >>> a = {'x': 1, 'y_1': {'a_2': 2}}
+    >>> normalize_nested_keys(a)  # doctest: +SKIP
```
Just curious, why is this doctest skipped?

jrbourbeau replied:
Good catch, thanks! I initially copied that example from another docstring in the `config.py` module and forgot to remove `# doctest: +SKIP`. Just pushed a commit with it removed.

mrocklin commented on dask/config.py:
```diff
@@ -102,6 +102,36 @@ def merge(*dicts):
+    for key, value in config.items():
+        if isinstance(value, Mapping):
+            value = normalize_nested_keys(value)
+        key_norm = normalize_key(key)
```
It looks like `{1: 2}` is valid yaml so not all keys are strings.

mrocklin replied:
I don't know when this would come up, but it's not entirely out of scope.

jrbourbeau replied:
Thanks for the review, I added a check to only call the `replace` method if the key is a string

mrocklin commented:
A small nit, but otherwise I'm +1.

mrocklin commented:
Thanks @jrbourbeau !

jcrist commented:
This is causing issues when keys aren't really config keys, but are things like environment variable names. E.g:

```
yarn:
  worker:
    env:
      MY_ENVIRONMENT_VARIABLE: foo
```

`MY_ENVIRONMENT_VARIABLE` gets rewritten as `MY-ENVIRONMENT-VARIABLE` on load, which breaks things.

I don't have a good suggestion for how to fix this. One thought would be to have modules register a set of known keys (e.g. `yarn.worker.env`), and only normalize those.

As an aside, I've very much enjoyed my experience hacking on Jupyter, their traitlets based configuration system (where config files are just python) has some nice benefits. I'm not suggesting we switch (that'd be hard to do), but having configuration schemas has its benefits.

djhoese commented:
@jcrist This is the same kind of issue in https://github.com/dask/dask/issues/4366 where the keys are actually dicts or values that need to be used as-is. As I suggested there, I think the "easy" solution that will work for all cases I've seen so far is to use an underscore as the normalized value. Then we can put off the proper fix for this until someone *needs* hyphenated keys.

jcrist commented:
That would be an easy band-aid, but I still think we shouldn't normalize anything that isn't explicitly a configuration key.
